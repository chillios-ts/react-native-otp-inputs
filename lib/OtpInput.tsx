import React, { ReactElement, Component } from 'react'
import { TextInput, View, StyleSheet } from 'react-native'

interface Props {
  containerStyles?: any
  error?: boolean
  focusedBorderColor?: string
  handleBackspace: (event) => void
  inputStyles?: any
  textErrorColor?: string
  unFocusedBorderColor: string
  updateText: (text) => void
  value?: string
}

interface State {
  isFocused: boolean
}

export default class OtpInput extends Component<Props, State> {
  state = {
    isFocused: false,
  }

  private input: ReactElement<TextInput>

  public render() {
    const {
      containerStyles,
      error,
      focusedBorderColor,
      handleBackspace,
      inputStyles,
      textErrorColor,
      unFocusedBorderColor,
      updateText,
      value,
    } = this.props

    return (
      <View
        style={[
          defaultStyles.container,
          containerStyles,
          { borderColor: this.state.isFocused ? focusedBorderColor : unFocusedBorderColor },
        ]}
      >
        <TextInput
          clearTextOnFocus={true}
          keyboardType="phone-pad"
          maxLength={1}
          onBlur={this._onBlur}
          onChangeText={updateText}
          onFocus={this._onFocus}
          onKeyPress={handleBackspace}
          ref={input => (this.input = input as any)}
          selectTextOnFocus={true}
          style={[defaultStyles.otpInput, inputStyles, error && { color: textErrorColor }]}
          underlineColorAndroid="transparent"
          value={value}
        />
      </View>
    )
  }

  private _onFocus = () => this.setState({ isFocused: true })
  private _onBlur = () => this.setState({ isFocused: false })
}

const defaultStyles = StyleSheet.create({
  container: {
    backgroundColor: '#a1a1a1',
    borderRadius: 6,
    borderWidth: 1,
    height: 53,
    margin: 10,
  },
  otpInput: {
    color: '#ffffff',
    fontSize: 24,
    paddingTop: 10,
    textAlign: 'center',
    width: 40,
  },
})
