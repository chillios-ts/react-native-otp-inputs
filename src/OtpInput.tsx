import React, { useState, useEffect, forwardRef } from 'react';
import {
  Platform,
  StyleProp,
  TextInput,
  TextStyle,
  View,
  ViewStyle,
  TextInputProps,
} from 'react-native';

type Props = TextInputProps & {
  inputContainerStyles?: StyleProp<ViewStyle>;
  firstInput: boolean;
  focusStyles?: StyleProp<ViewStyle>;
  inputStyles?: StyleProp<TextStyle>;
  numberOfInputs: number;
  handleTextChange: (text: string) => void;
};

const majorVersionIOS: number = parseInt(`${Platform.Version}`, 10);
const isOTPSupported: boolean = Platform.OS === 'ios' && majorVersionIOS >= 12;

const OtpInput = forwardRef<TextInput, Props>(
  (
    {
      firstInput,
      focusStyles,
      handleTextChange,
      inputContainerStyles,
      inputStyles,
      numberOfInputs,
      placeholder,
      selectTextOnFocus,
      value,
      ...rest
    },
    ref,
  ) => {
    const [focused, setFocused] = useState(false);

    useEffect(() => {
      // @ts-ignore
      ref.current.setNativeProps({ value, text: value });
    }, [ref, value]);

    return (
      <View style={[inputContainerStyles, focused && focusStyles]}>
        <TextInput
          maxLength={firstInput ? numberOfInputs : 1}
          onBlur={() => setFocused(false)}
          onChangeText={handleTextChange}
          onFocus={() => setFocused(true)}
          ref={ref}
          placeholder={placeholder}
          // https://github.com/facebook/react-native/issues/18339
          selectTextOnFocus={Platform.select({ ios: selectTextOnFocus, android: true })}
          style={inputStyles}
          textContentType={isOTPSupported ? 'oneTimeCode' : 'none'}
          underlineColorAndroid="transparent"
          {...rest}
        />
      </View>
    );
  },
);

export default OtpInput;
