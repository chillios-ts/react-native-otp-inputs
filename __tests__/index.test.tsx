import { Keyboard } from 'react-native'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

import OtpInputs from '../src/'
import OtpInput from '../src/OtpInput'

Keyboard.dismiss = jest.fn()

describe('<OtpInputs />', () => {
  const wrapper = renderer.create(<OtpInputs />)

  test('render', done => {
    expect(wrapper.toJSON()).toMatchSnapshot()
    expect(wrapper.root.findAllByType(OtpInput).length).toEqual(4)
    const sixInputsWrapper = renderer.create(<OtpInputs numberOfInputs={6} />)
    expect(sixInputsWrapper.root.findAllByType(OtpInput).length).toEqual(6)
    done()
  })

  test('rendering with error', () => {
    const testWrapper = renderer.create(<OtpInputs errorMessage="ExampleErrorMessage" />)
    const errorTextElement = testWrapper.root.findByProps({ testID: 'errorText' })

    expect(testWrapper).toMatchSnapshot()
    expect(errorTextElement._fiber.stateNode.props.children).toEqual('ExampleErrorMessage')
  })

  describe('_updateText', () => {
    describe('when input is filled in', () => {
      test('should call _focusInput and not call `Keyboard.dismiss`', () => {
        const fakeEvent = { nativeEvent: { text: '1' } }
        const inputIndex = 0
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._focusInput = jest.fn()
        wrapperInstance._updateText(fakeEvent, inputIndex)

        expect(wrapperInstance._focusInput).toBeCalled()
        expect(Keyboard.dismiss).not.toBeCalled()
      })
    })

    describe('when no text is given', () => {
      test('should not any function', () => {
        const fakeEvent = { nativeEvent: { text: '' } }
        const inputIndex = 2
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._focusInput = jest.fn()
        wrapperInstance._updateText(fakeEvent, inputIndex)

        expect(wrapperInstance._focusInput).not.toBeCalled()
        expect(Keyboard.dismiss).not.toBeCalled()
      })
    })

    describe('when input that does not exist is filled in (fast typing)', () => {
      test('should not any function', () => {
        const fakeEvent = { nativeEvent: { text: '1' } }
        const inputIndex = 5
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._focusInput = jest.fn()
        wrapperInstance._updateText(fakeEvent, inputIndex)

        expect(wrapperInstance._focusInput).not.toBeCalled()
        expect(Keyboard.dismiss).not.toBeCalled()
      })
    })

    describe('when last input is filled in', () => {
      test('should not call _focusInput and call `Keyboard.dismiss`', () => {
        const fakeEvent = { nativeEvent: { text: '1' } }
        const inputIndex = 3
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._focusInput = jest.fn()
        wrapperInstance._updateText(fakeEvent, inputIndex)

        expect(wrapperInstance._focusInput).not.toBeCalled()
        expect(Keyboard.dismiss).toBeCalled()
      })
    })

    describe('when input with value is filled with new one', () => {
      test('should override first value', () => {
        const wrapper = renderer.create(<OtpInputs />)

        const fakeEvent = { nativeEvent: { text: '12' } }
        const inputIndex = 0
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._updateText(fakeEvent, inputIndex)

        expect(wrapperInstance.state.otpCode.join('')).toEqual('2')
        expect(Keyboard.dismiss).toBeCalled()
      })
    })
  })

  describe('_handleBackspace', () => {
    describe('when first input is cleared', () => {
      test('should not call _focusInput', () => {
        const fakeNativeEvent = { nativeEvent: { key: 'Backspace' } }
        const inputIndex = 0
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._focusInput = jest.fn()
        wrapperInstance._handleBackspace(fakeNativeEvent, inputIndex)

        expect(wrapperInstance._focusInput).not.toBeCalled()
      })
    })

    describe('when key is not backspace', () => {
      test('should not call _focusInput', () => {
        const fakeNativeEvent = { nativeEvent: { key: 'Enter' } }
        const inputIndex = 0
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._focusInput = jest.fn()
        wrapperInstance._handleBackspace(fakeNativeEvent, inputIndex)

        expect(wrapperInstance._focusInput).not.toBeCalled()
      })
    })

    describe('when index is equal last index', () => {
      test('should call _focusInput', () => {
        const fakeNativeEvent = { nativeEvent: { key: 'Backspace' } }
        const inputIndex = 3
        const wrapperInstance = wrapper.getInstance()
        wrapperInstance._focusInput = jest.fn()
        wrapperInstance._handleBackspace(fakeNativeEvent, inputIndex)

        expect(wrapperInstance._focusInput).toBeCalled()
      })
    })
  })
})

describe('OtpInput events', () => {
  const wrapper = renderer.create(<OtpInputs />)

  describe('handleBackspace', () => {
    const wrapperInstance = wrapper.getInstance()
    wrapperInstance._handleBackspace = jest.fn()
    const OtpInputComponent = wrapper.root.findAllByType(OtpInput)[0]

    OtpInputComponent.props.handleBackspace()
    expect(wrapperInstance._handleBackspace).toBeCalled()
  })

  describe('updateText', () => {
    const wrapperInstance = wrapper.getInstance()
    wrapperInstance._updateText = jest.fn()
    const OtpInputComponent = wrapper.root.findAllByType(OtpInput)[0]

    OtpInputComponent.props.updateText()
    expect(wrapperInstance._updateText).toBeCalled()
  })
})
