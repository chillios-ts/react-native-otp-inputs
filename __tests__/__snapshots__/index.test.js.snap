// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OtpInputs /> render 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            false,
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            false,
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            false,
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            false,
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;

exports[`<OtpInputs /> rendering with error 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "marginHorizontal": 25,
        },
        undefined,
      ]
    }
  >
    <Text
      testID="errorText"
    >
      ExampleErrorMessage
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            Object {
              "color": "#ff0000",
            },
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            Object {
              "color": "#ff0000",
            },
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            Object {
              "color": "#ff0000",
            },
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "height": 53,
            "margin": 10,
          },
          undefined,
          Object {
            "borderColor": "#a0a0a0",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        keyboardType="phone-pad"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={false}
        selectTextOnFocus={true}
        style={
          Array [
            Object {
              "fontSize": 24,
              "paddingTop": 10,
              "textAlign": "center",
              "width": 40,
            },
            undefined,
            Object {
              "color": "#ff0000",
            },
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;
